### DISCUSSION TITLE
Recommended way to parse error informations from failing request?

### DISCUSSION THREAD

**Discussion by jsiedentop** [ory/kratos#1144] [Q&A] [answered]
I use the Login Flow with Username/Email Password Method. When no data for username/password is transfered I got this error-response (400 - Bad Request):
```
{
	"id":"e8834d5b-dc91-415b-ade9-cbfde215c40a",
	"type":"api",
	"expires_at":"2021-03-12T19:43:52.076074Z",
	"issued_at":"2021-03-12T19:33:52.076074Z",
	"request_url":"http://10.0.2.2:4433/self-service/login/api",
	"active":"password",
	"messages":null,
	"methods":{
		"password":{
			"method":"password",
			"config":{
				"action":"http://127.0.0.1:4433/self-service/login/methods/password?flow=e8834d5b-dc91-415b-ade9-cbfde215c40a",
				"method":"POST",
				"fields":[
					{"name":"identifier","type":"text","required":true,"value":"","messages":[{"id":4000001,"text":"length must be \u003e= 1, but got 0","type":"error"}]},
					{"name":"password","type":"password","required":true,"messages":[{"id":4000001,"text":"length must be \u003e= 1, but got 0","type":"error"}]},
					{"name":"csrf_token","type":"hidden","required":true}
				]
			}
		}
	},
	"forced":false
}
```

In the documentation I can see, that the data of this response is structured with the loginFlow schema:

https://www.ory.sh/kratos/docs/reference/api/#complete-login-flow-with-usernameemail-password-method
https://www.ory.sh/kratos/docs/reference/api/#schemaloginflow

The "methods" property is documented as followed: 
```
Name: methods
Type: object	
Required: true	
Restrictions: none	
Description: List of login methods This is the list of available login methods with their required form fields, such as identifier and password for the password login method. This will also contain error messages such as "password can not be empty".
```
It seems, that there is no defined structure for the containing data (type of "methods" property is object). So giving the user a hint, why the login failed is very hard, because the structure is not documented and it is possible, that the response for different error cases are structured in different ways. The sdk offers only an ApiException with all the content as a string property.

Is there any recommended way to parse the error information?

**Comment by Benehiko** [comment, marked as answer]
Hi @jsiedentop,

Thank you for reaching out :smile: 

This documentation might help you with regards to error messages:
- https://www.ory.sh/kratos/docs/concepts/ui-user-interface#messages
- https://www.ory.sh/kratos/docs/self-service/flows/user-facing-errors/
- https://www.ory.sh/kratos/docs/reference/api/#get-user-facing-self-service-errors

Please let me know if this resolves your problem :smiley:
